fix initial e2e-test failure in make test-e2e

fix initial e2e-test failure in make test-e2e
~~~
[akaris@linux ingress-node-firewall (sdn-3315)]$ make test-e2e
/home/akaris/development/ingress-node-firewall/bin/controller-gen object:headerFile="hack/boilerplate.go.txt" paths="./..."
go fmt ./...
go vet ./...
==== Generating DaemonSet manifest
hack/generate-daemon-manifest.sh
/home/akaris/development/ingress-node-firewall/bin/controller-gen rbac:roleName=manager-role crd webhook paths="./..." output:crd:artifacts:config=config/crd/bases
rm -rf /tmp/test_validation_logs/
mkdir -p /tmp/test_validation_logs/
go test --tags=validationtests -v ./test/e2e/validation -ginkgo.v -junit /tmp/test_validation_logs/ -report /tmp/test_validation_logs/ 
=== RUN   TestValidation
Running Suite: Ingress Node Firewall Operator Validation Suite
==============================================================
Random Seed: 1660219716
Will run 4 of 4 specs

IngressNodeFirewall IngressNodeFirewall 
  should have the IngressNodeFirewall Operator deployment in running state
  /home/akaris/development/ingress-node-firewall/test/e2e/validation/tests/validation.go:30
•
------------------------------
IngressNodeFirewall IngressNodeFirewall 
  should have the IngressNodeFirewallConfig CRD available in the cluster
  /home/akaris/development/ingress-node-firewall/test/e2e/validation/tests/validation.go:52
•
------------------------------
IngressNodeFirewall IngressNodeFirewall 
  should have the IngressNodeFirewall CRD available in the cluster
  /home/akaris/development/ingress-node-firewall/test/e2e/validation/tests/validation.go:58
•
------------------------------
IngressNodeFirewall IngressNodeFirewall 
  should have the IngressNodeFirewallNodeState CRD available in the cluster
  /home/akaris/development/ingress-node-firewall/test/e2e/validation/tests/validation.go:64
•
JUnit report was created: /tmp/test_validation_logs/validation_junit.xml

Ran 4 of 4 Specs in 0.039 seconds
SUCCESS! -- 4 Passed | 0 Failed | 0 Pending | 0 Skipped

You're using deprecated Ginkgo functionality:
=============================================
Ginkgo 2.0 is under active development and will introduce several new features, improvements, and a small handful of breaking changes.
A release candidate for 2.0 is now available and 2.0 should GA in Fall 2021.  Please give the RC a try and send us feedback!
  - To learn more, view the migration guide at https://github.com/onsi/ginkgo/blob/ver2/docs/MIGRATING_TO_V2.md
  - For instructions on using the Release Candidate visit https://github.com/onsi/ginkgo/blob/ver2/docs/MIGRATING_TO_V2.md#using-the-beta
  - To comment, chime in at https://github.com/onsi/ginkgo/issues/711

  You are using a custom reporter.  Support for custom reporters will likely be removed in V2.  Most users were using them to generate junit or teamcity reports and this functionality will be merged into the core reporter.  In addition, Ginkgo 2.0 will support emitting a JSON-formatted report that users can then manipulate to generate custom reports.

  If this change will be impactful to you please leave a comment on https://github.com/onsi/ginkgo/issues/711
  Learn more at: https://github.com/onsi/ginkgo/blob/ver2/docs/MIGRATING_TO_V2.md#removed-custom-reporters

To silence deprecations that can be silenced set the following environment variable:
  ACK_GINKGO_DEPRECATIONS=1.16.5

--- PASS: TestValidation (0.07s)
PASS
ok  	github.com/openshift/ingress-node-firewall/test/e2e/validation	0.186s
rm -rf /tmp/test_e2e_logs/
mkdir -p /tmp/test_e2e_logs/
go test --tags=e2etests -v ./test/e2e/functional -ginkgo.v -junit /tmp/test_e2e_logs/ -report /tmp/test_e2e_logs/ 
=== RUN   TestE2E
Running Suite: Ingress Node Firewall Operator E2E Suite
=======================================================
Random Seed: 1660219719
Will run 16 of 16 specs

Ingress Node Firewall IngressNodeFirewall deploy 
  should run Ingress node firewall DaemonSet
  /home/akaris/development/ingress-node-firewall/test/e2e/functional/tests/e2e.go:70
STEP: checking Ingress node firewall daemonset is in running state

• [SLOW TEST:8.147 seconds]
Ingress Node Firewall
/home/akaris/development/ingress-node-firewall/test/e2e/functional/tests/e2e.go:38
  IngressNodeFirewall deploy
  /home/akaris/development/ingress-node-firewall/test/e2e/functional/tests/e2e.go:39
    should run Ingress node firewall DaemonSet
    /home/akaris/development/ingress-node-firewall/test/e2e/functional/tests/e2e.go:70
------------------------------
Ingress Node Firewall IngressNodeFirewall deploy 
  should run Ingress node firewall apply rules and check the actions
  /home/akaris/development/ingress-node-firewall/test/e2e/functional/tests/e2e.go:94
STEP: get nodes IP addresses with matching labels and ping their IPs
STEP: creating ingress node firewall rules
STEP: checking Ingress node firewall rules resource is created
STEP: checking ingress node firewall nodeState resource is created
STEP: checking Ingress node firewall rules deny ping packets
STEP: checking ingress node firewall events are generated
STEP: checking Ingress node firewall nodeState resource is deleted
STEP: checking Ingress node firewall rules resource is deleted

• [SLOW TEST:32.421 seconds]
Ingress Node Firewall
/home/akaris/development/ingress-node-firewall/test/e2e/functional/tests/e2e.go:38
  IngressNodeFirewall deploy
  /home/akaris/development/ingress-node-firewall/test/e2e/functional/tests/e2e.go:39
    should run Ingress node firewall apply rules and check the actions
    /home/akaris/development/ingress-node-firewall/test/e2e/functional/tests/e2e.go:94
------------------------------
Ingress Node Firewall Statistics 
  should expose at least one endpoint via a daemon metrics service
  /home/akaris/development/ingress-node-firewall/test/e2e/functional/tests/e2e.go:258

• [SLOW TEST:5.059 seconds]
Ingress Node Firewall
/home/akaris/development/ingress-node-firewall/test/e2e/functional/tests/e2e.go:38
  Statistics
  /home/akaris/development/ingress-node-firewall/test/e2e/functional/tests/e2e.go:227
    should expose at least one endpoint via a daemon metrics service
    /home/akaris/development/ingress-node-firewall/test/e2e/functional/tests/e2e.go:258
------------------------------
Ingress Node Firewall Statistics 
  should expose daemon metrics
  /home/akaris/development/ingress-node-firewall/test/e2e/functional/tests/e2e.go:285

• [SLOW TEST:39.396 seconds]
Ingress Node Firewall
/home/akaris/development/ingress-node-firewall/test/e2e/functional/tests/e2e.go:38
  Statistics
  /home/akaris/development/ingress-node-firewall/test/e2e/functional/tests/e2e.go:227
    should expose daemon metrics
    /home/akaris/development/ingress-node-firewall/test/e2e/functional/tests/e2e.go:285
------------------------------
Ingress Node Firewall Webhook 
  should allow valid ingressnodefirewall TCP rule
  /home/akaris/development/ingress-node-firewall/test/e2e/functional/tests/e2e.go:425
•
------------------------------
Ingress Node Firewall Webhook 
  should allow valid ingressnodefirewall UDP rule
  /home/akaris/development/ingress-node-firewall/test/e2e/functional/tests/e2e.go:434
•
------------------------------
Ingress Node Firewall Webhook 
  should allow valid ingressnodefirewall ICMPV4 rule
  /home/akaris/development/ingress-node-firewall/test/e2e/functional/tests/e2e.go:443
•
------------------------------
Ingress Node Firewall Webhook 
  should allow valid ingressnodefirewall ICMPV6 rule
  /home/akaris/development/ingress-node-firewall/test/e2e/functional/tests/e2e.go:452
•
------------------------------
Ingress Node Firewall Webhook 
  should allow valid ingressnodefirewall SCTP rule
  /home/akaris/development/ingress-node-firewall/test/e2e/functional/tests/e2e.go:461
•
------------------------------
Ingress Node Firewall Webhook 
  should block any rules which conflict with failsafe rules
  /home/akaris/development/ingress-node-firewall/test/e2e/functional/tests/e2e.go:470
•
------------------------------
Ingress Node Firewall Webhook 
  should block invalid order
  /home/akaris/development/ingress-node-firewall/test/e2e/functional/tests/e2e.go:487

• Failure in Spec Teardown (AfterEach) [0.012 seconds]
Ingress Node Firewall
/home/akaris/development/ingress-node-firewall/test/e2e/functional/tests/e2e.go:38
  Webhook
  /home/akaris/development/ingress-node-firewall/test/e2e/functional/tests/e2e.go:394
    should block invalid order [AfterEach]
    /home/akaris/development/ingress-node-firewall/test/e2e/functional/tests/e2e.go:487

    Unexpected error:
        <*errors.StatusError | 0xc0005cc780>: {
            ErrStatus: {
                TypeMeta: {Kind: "", APIVersion: ""},
                ListMeta: {
                    SelfLink: "",
                    ResourceVersion: "",
                    Continue: "",
                    RemainingItemCount: nil,
                },
                Status: "Failure",
                Message: "daemonsets.apps \"ingress-node-firewall-daemon\" not found",
                Reason: "NotFound",
                Details: {
                    Name: "ingress-node-firewall-daemon",
                    Group: "apps",
                    Kind: "daemonsets",
                    UID: "",
                    Causes: nil,
                    RetryAfterSeconds: 0,
                },
                Code: 404,
            },
        }
        daemonsets.apps "ingress-node-firewall-daemon" not found
    occurred

    /home/akaris/development/ingress-node-firewall/test/e2e/functional/tests/e2e.go:417
------------------------------
Ingress Node Firewall Webhook 
  should block TCP rule ICMP defined
  /home/akaris/development/ingress-node-firewall/test/e2e/functional/tests/e2e.go:497
•
------------------------------
Ingress Node Firewall Webhook 
  should block UDP rule ICMP defined
  /home/akaris/development/ingress-node-firewall/test/e2e/functional/tests/e2e.go:506
•
------------------------------
Ingress Node Firewall Webhook 
  should block SCTP rule ICMP defined
  /home/akaris/development/ingress-node-firewall/test/e2e/functional/tests/e2e.go:515
•
------------------------------
Ingress Node Firewall Webhook 
  should block ICMPV4 with TCP port defined
  /home/akaris/development/ingress-node-firewall/test/e2e/functional/tests/e2e.go:524
•
------------------------------
Ingress Node Firewall Webhook 
  should block ICMPV6 with TCP port defined
  /home/akaris/development/ingress-node-firewall/test/e2e/functional/tests/e2e.go:533
•
JUnit report was created: /tmp/test_e2e_logs/e2e_junit.xml


Summarizing 1 Failure:

[Fail] Ingress Node Firewall Webhook [AfterEach] should block invalid order 
/home/akaris/development/ingress-node-firewall/test/e2e/functional/tests/e2e.go:417

Ran 16 of 16 Specs in 109.838 seconds
FAIL! -- 15 Passed | 1 Failed | 0 Pending | 0 Skipped

You're using deprecated Ginkgo functionality:
=============================================
Ginkgo 2.0 is under active development and will introduce several new features, improvements, and a small handful of breaking changes.
A release candidate for 2.0 is now available and 2.0 should GA in Fall 2021.  Please give the RC a try and send us feedback!
  - To learn more, view the migration guide at https://github.com/onsi/ginkgo/blob/ver2/docs/MIGRATING_TO_V2.md
  - For instructions on using the Release Candidate visit https://github.com/onsi/ginkgo/blob/ver2/docs/MIGRATING_TO_V2.md#using-the-beta
  - To comment, chime in at https://github.com/onsi/ginkgo/issues/711

  You are using a custom reporter.  Support for custom reporters will likely be removed in V2.  Most users were using them to generate junit or teamcity reports and this functionality will be merged into the core reporter.  In addition, Ginkgo 2.0 will support emitting a JSON-formatted report that users can then manipulate to generate custom reports.

  If this change will be impactful to you please leave a comment on https://github.com/onsi/ginkgo/issues/711
  Learn more at: https://github.com/onsi/ginkgo/blob/ver2/docs/MIGRATING_TO_V2.md#removed-custom-reporters

To silence deprecations that can be silenced set the following environment variable:
  ACK_GINKGO_DEPRECATIONS=1.16.5

--- FAIL: TestE2E (109.86s)
FAIL
FAIL	github.com/openshift/ingress-node-firewall/test/e2e/functional	109.938s
FAIL
make: *** [Makefile:134: test-functional] Error 1
[akaris@linux ingress-node-firewall (sdn-3315)]$ make test-e2e
/home/akaris/development/ingress-node-firewall/bin/controller-gen object:headerFile="hack/boilerplate.go.txt" paths="./..."
go fmt ./...
go vet ./...
==== Generating DaemonSet manifest
hack/generate-daemon-manifest.sh
/home/akaris/development/ingress-node-firewall/bin/controller-gen rbac:roleName=manager-role crd webhook paths="./..." output:crd:artifacts:config=config/crd/bases
rm -rf /tmp/test_validation_logs/
mkdir -p /tmp/test_validation_logs/
go test --tags=validationtests -v ./test/e2e/validation -ginkgo.v -junit /tmp/test_validation_logs/ -report /tmp/test_validation_logs/ 
=== RUN   TestValidation
Running Suite: Ingress Node Firewall Operator Validation Suite
==============================================================
Random Seed: 1660219843
Will run 4 of 4 specs

IngressNodeFirewall IngressNodeFirewall 
  should have the IngressNodeFirewall Operator deployment in running state
  /home/akaris/development/ingress-node-firewall/test/e2e/validation/tests/validation.go:30
•
------------------------------
IngressNodeFirewall IngressNodeFirewall 
  should have the IngressNodeFirewallConfig CRD available in the cluster
  /home/akaris/development/ingress-node-firewall/test/e2e/validation/tests/validation.go:52
•
------------------------------
IngressNodeFirewall IngressNodeFirewall 
  should have the IngressNodeFirewall CRD available in the cluster
  /home/akaris/development/ingress-node-firewall/test/e2e/validation/tests/validation.go:58
•
------------------------------
IngressNodeFirewall IngressNodeFirewall 
  should have the IngressNodeFirewallNodeState CRD available in the cluster
  /home/akaris/development/ingress-node-firewall/test/e2e/validation/tests/validation.go:64
•
JUnit report was created: /tmp/test_validation_logs/validation_junit.xml

Ran 4 of 4 Specs in 0.040 seconds
SUCCESS! -- 4 Passed | 0 Failed | 0 Pending | 0 Skipped

You're using deprecated Ginkgo functionality:
=============================================
Ginkgo 2.0 is under active development and will introduce several new features, improvements, and a small handful of breaking changes.
A release candidate for 2.0 is now available and 2.0 should GA in Fall 2021.  Please give the RC a try and send us feedback!
  - To learn more, view the migration guide at https://github.com/onsi/ginkgo/blob/ver2/docs/MIGRATING_TO_V2.md
  - For instructions on using the Release Candidate visit https://github.com/onsi/ginkgo/blob/ver2/docs/MIGRATING_TO_V2.md#using-the-beta
  - To comment, chime in at https://github.com/onsi/ginkgo/issues/711

  You are using a custom reporter.  Support for custom reporters will likely be removed in V2.  Most users were using them to generate junit or teamcity reports and this functionality will be merged into the core reporter.  In addition, Ginkgo 2.0 will support emitting a JSON-formatted report that users can then manipulate to generate custom reports.

  If this change will be impactful to you please leave a comment on https://github.com/onsi/ginkgo/issues/711
  Learn more at: https://github.com/onsi/ginkgo/blob/ver2/docs/MIGRATING_TO_V2.md#removed-custom-reporters

To silence deprecations that can be silenced set the following environment variable:
  ACK_GINKGO_DEPRECATIONS=1.16.5

--- PASS: TestValidation (0.06s)
PASS
ok  	github.com/openshift/ingress-node-firewall/test/e2e/validation	0.114s
rm -rf /tmp/test_e2e_logs/
mkdir -p /tmp/test_e2e_logs/
go test --tags=e2etests -v ./test/e2e/functional -ginkgo.v -junit /tmp/test_e2e_logs/ -report /tmp/test_e2e_logs/ 
=== RUN   TestE2E
Running Suite: Ingress Node Firewall Operator E2E Suite
=======================================================
Random Seed: 1660219845
Will run 16 of 16 specs

Ingress Node Firewall IngressNodeFirewall deploy 
  should run Ingress node firewall DaemonSet
  /home/akaris/development/ingress-node-firewall/test/e2e/functional/tests/e2e.go:70
STEP: checking Ingress node firewall daemonset is in running state
•
------------------------------
Ingress Node Firewall IngressNodeFirewall deploy 
  should run Ingress node firewall apply rules and check the actions
  /home/akaris/development/ingress-node-firewall/test/e2e/functional/tests/e2e.go:94
STEP: get nodes IP addresses with matching labels and ping their IPs
STEP: creating ingress node firewall rules
STEP: checking Ingress node firewall rules resource is created
STEP: checking ingress node firewall nodeState resource is created
STEP: checking Ingress node firewall rules deny ping packets
STEP: checking ingress node firewall events are generated
STEP: checking Ingress node firewall nodeState resource is deleted
STEP: checking Ingress node firewall rules resource is deleted

• [SLOW TEST:28.268 seconds]
Ingress Node Firewall
/home/akaris/development/ingress-node-firewall/test/e2e/functional/tests/e2e.go:38
  IngressNodeFirewall deploy
  /home/akaris/development/ingress-node-firewall/test/e2e/functional/tests/e2e.go:39
    should run Ingress node firewall apply rules and check the actions
    /home/akaris/development/ingress-node-firewall/test/e2e/functional/tests/e2e.go:94
------------------------------
Ingress Node Firewall Statistics 
  should expose at least one endpoint via a daemon metrics service
  /home/akaris/development/ingress-node-firewall/test/e2e/functional/tests/e2e.go:258
•
------------------------------
Ingress Node Firewall Statistics 
  should expose daemon metrics
  /home/akaris/development/ingress-node-firewall/test/e2e/functional/tests/e2e.go:285

• [SLOW TEST:34.544 seconds]
Ingress Node Firewall
/home/akaris/development/ingress-node-firewall/test/e2e/functional/tests/e2e.go:38
  Statistics
  /home/akaris/development/ingress-node-firewall/test/e2e/functional/tests/e2e.go:227
    should expose daemon metrics
    /home/akaris/development/ingress-node-firewall/test/e2e/functional/tests/e2e.go:285
------------------------------
Ingress Node Firewall Webhook 
  should allow valid ingressnodefirewall TCP rule
  /home/akaris/development/ingress-node-firewall/test/e2e/functional/tests/e2e.go:425
•
------------------------------
Ingress Node Firewall Webhook 
  should allow valid ingressnodefirewall UDP rule
  /home/akaris/development/ingress-node-firewall/test/e2e/functional/tests/e2e.go:434
•
------------------------------
Ingress Node Firewall Webhook 
  should allow valid ingressnodefirewall ICMPV4 rule
  /home/akaris/development/ingress-node-firewall/test/e2e/functional/tests/e2e.go:443
•
------------------------------
Ingress Node Firewall Webhook 
  should allow valid ingressnodefirewall ICMPV6 rule
  /home/akaris/development/ingress-node-firewall/test/e2e/functional/tests/e2e.go:452
•
------------------------------
Ingress Node Firewall Webhook 
  should allow valid ingressnodefirewall SCTP rule
  /home/akaris/development/ingress-node-firewall/test/e2e/functional/tests/e2e.go:461
•
------------------------------
Ingress Node Firewall Webhook 
  should block any rules which conflict with failsafe rules
  /home/akaris/development/ingress-node-firewall/test/e2e/functional/tests/e2e.go:470
•
------------------------------
Ingress Node Firewall Webhook 
  should block invalid order
  /home/akaris/development/ingress-node-firewall/test/e2e/functional/tests/e2e.go:487
•
------------------------------
Ingress Node Firewall Webhook 
  should block TCP rule ICMP defined
  /home/akaris/development/ingress-node-firewall/test/e2e/functional/tests/e2e.go:497
•
------------------------------
Ingress Node Firewall Webhook 
  should block UDP rule ICMP defined
  /home/akaris/development/ingress-node-firewall/test/e2e/functional/tests/e2e.go:506
•
------------------------------
Ingress Node Firewall Webhook 
  should block SCTP rule ICMP defined
  /home/akaris/development/ingress-node-firewall/test/e2e/functional/tests/e2e.go:515
•
------------------------------
Ingress Node Firewall Webhook 
  should block ICMPV4 with TCP port defined
  /home/akaris/development/ingress-node-firewall/test/e2e/functional/tests/e2e.go:524
•
------------------------------
Ingress Node Firewall Webhook 
  should block ICMPV6 with TCP port defined
  /home/akaris/development/ingress-node-firewall/test/e2e/functional/tests/e2e.go:533
•
JUnit report was created: /tmp/test_e2e_logs/e2e_junit.xml

Ran 16 of 16 Specs in 65.509 seconds
SUCCESS! -- 16 Passed | 0 Failed | 0 Pending | 0 Skipped

You're using deprecated Ginkgo functionality:
=============================================
Ginkgo 2.0 is under active development and will introduce several new features, improvements, and a small handful of breaking changes.
A release candidate for 2.0 is now available and 2.0 should GA in Fall 2021.  Please give the RC a try and send us feedback!
  - To learn more, view the migration guide at https://github.com/onsi/ginkgo/blob/ver2/docs/MIGRATING_TO_V2.md
  - For instructions on using the Release Candidate visit https://github.com/onsi/ginkgo/blob/ver2/docs/MIGRATING_TO_V2.md#using-the-beta
  - To comment, chime in at https://github.com/onsi/ginkgo/issues/711

  You are using a custom reporter.  Support for custom reporters will likely be removed in V2.  Most users were using them to generate junit or teamcity reports and this functionality will be merged into the core reporter.  In addition, Ginkgo 2.0 will support emitting a JSON-formatted report that users can then manipulate to generate custom reports.

  If this change will be impactful to you please leave a comment on https://github.com/onsi/ginkgo/issues/711
  Learn more at: https://github.com/onsi/ginkgo/blob/ver2/docs/MIGRATING_TO_V2.md#removed-custom-reporters

To silence deprecations that can be silenced set the following environment variable:
  ACK_GINKGO_DEPRECATIONS=1.16.5

--- PASS: TestE2E (65.53s)
PASS
ok  	github.com/openshift/ingress-node-firewall/test/e2e/functional	65.603s
~~~

--- a/controllers/ingressnodefirewall_controller.go
+++ b/controllers/ingressnodefirewall_controller.go
@@ -23,7 +23,6 @@ import (
 	"github.com/go-logr/logr"
 
 	infv1alpha1 "github.com/openshift/ingress-node-firewall/api/v1alpha1"
-
 	v1 "k8s.io/api/core/v1"
 	"k8s.io/apimachinery/pkg/api/equality"
 	"k8s.io/apimachinery/pkg/api/errors"
@@ -211,7 +210,7 @@ func (r *IngressNodeFirewallReconciler) triggerReconciliation(object client.Obje
 		return []reconcile.Request{}
 	}
 
-	// We do not need to reconcile anything if there are no items of type IngressNodeFirwall.
+	// We do not need to reconcile anything if there are no items of type IngressNodeFirewall.
 	if len(ingressNodeFirewallList.Items) == 0 {
 		return []reconcile.Request{}
 	}
@@ -324,17 +323,6 @@ func (r *IngressNodeFirewallReconciler) buildNodeStates(
 				continue withNextNode
 			}
 			for _, iface := range firewallObj.Spec.Interfaces {
-				// Verify if an invalid interface name was specified.
-				// On error, report the error in the status field and continue with the next node.
-				if !isLegalInterfaceName(iface) {
-					state.Status = infv1alpha1.IngressNodeFirewallNodeStateStatus{
-						SyncStatus:       infv1alpha1.SyncError,
-						SyncErrorMessage: fmt.Sprintf("Invalid interface name %s", iface),
-					}
-					// Write back the state to the map and then continue with the next node.
-					nodeStates[node.Name] = state
-					continue withNextNode
-				}
 				// Create the rules for the node spec if they do not yet exist for this interface.
 				if _, ok := state.Spec.InterfaceIngressRules[iface]; !ok {
 					state.Spec.InterfaceIngressRules[iface] = []infv1alpha1.IngressNodeFirewallRules{}
@@ -359,13 +347,6 @@ func (r *IngressNodeFirewallReconciler) buildNodeStates(
 	}
 
 	return nodeStates, nil
-}
-
-func isLegalInterfaceName(iface string) bool {
-	if iface == "" {
-		return false
-	}
-	return true
 }
 
 // mergeRuleSet merges 2 rulesets of type []infv1alpha1.IngressNodeFirewallRules.

--- a/controllers/ingressnodefirewall_controller_test.go
+++ b/controllers/ingressnodefirewall_controller_test.go
@@ -18,13 +18,12 @@ import (
 	"sigs.k8s.io/controller-runtime/pkg/client"
 )
 
+const (
+	timeout  = time.Second * 10
+	interval = time.Millisecond * 250
+)
+
 var _ = Describe("IngressNodeFirewall controller", func() {
-
-	const (
-		timeout  = time.Second * 10
-		interval = time.Millisecond * 250
-	)
-
 	ctx := context.Background()
 
 	ingressNodeFirewallName := "firewall1"

--- a/pkg/ebpfsyncer/ebpfsyncer.go
+++ b/pkg/ebpfsyncer/ebpfsyncer.go
@@ -6,16 +6,21 @@ import (
 	"net"
 	"sync"
 
-	"github.com/go-logr/logr"
 	infv1alpha1 "github.com/openshift/ingress-node-firewall/api/v1alpha1"
 	nodefwloader "github.com/openshift/ingress-node-firewall/pkg/ebpf"
 	"github.com/openshift/ingress-node-firewall/pkg/failsaferules"
+	intfs "github.com/openshift/ingress-node-firewall/pkg/interfaces"
 	"github.com/openshift/ingress-node-firewall/pkg/metrics"
+
+	"github.com/go-logr/logr"
 	"k8s.io/apimachinery/pkg/util/intstr"
 )
 
-var once sync.Once
-var instance EbpfSyncer
+var (
+	once                         sync.Once
+	instance                     EbpfSyncer
+	isValidInterfaceNameAndState = intfs.IsValidInterfaceNameAndState
+)
 
 // ebpfDaemon is a single point of contact that all reconciliation requests will send their desired state of
 // interface rules to. On the other side, ebpfDaemon makes sure that rules are attached and detached from / to the
@@ -112,6 +117,9 @@ func (e *ebpfSingleton) SyncInterfaceIngressRules(
 	// Also delete the rules associated to these interfaces. See TODO below.
 	for intf := range e.managedInterfaces {
 		logger.Info("Running detach operation for interface", "intf", intf)
+		if !isValidInterfaceNameAndState(intf) {
+			return fmt.Errorf("Fail to attach ingress node Firewall rules: invalid inteface %s", intf)
+		}
 		if _, ok := ifaceIngressRules[intf]; !ok {
 			_, err := e.c.IngressNodeFwAttach([]string{intf}, true)
 			if err != nil {

--- a/pkg/ebpfsyncer/ebpfsyncer_test.go
+++ b/pkg/ebpfsyncer/ebpfsyncer_test.go
@@ -13,6 +13,7 @@ import (
 	"time"
 
 	infv1alpha1 "github.com/openshift/ingress-node-firewall/api/v1alpha1"
+
 	"github.com/vishvananda/netlink"
 	"k8s.io/apimachinery/pkg/util/intstr"
 	"sigs.k8s.io/controller-runtime/pkg/log/zap"
@@ -37,7 +38,9 @@ func TestSyncInterfaceIngressRulesConnectDenyRule(t *testing.T) {
 
 	defer afterEach(t)
 	beforeEach(t)
-
+	isValidInterfaceNameAndState = func(ifName string) bool {
+		return true
+	}
 	rules := map[string][]infv1alpha1.IngressNodeFirewallRules{
 		fmt.Sprintf("%s0", interfacePrefix): {
 			{
@@ -96,7 +99,9 @@ func TestSyncInterfaceIngressRulesConnectAllowRule(t *testing.T) {
 
 	defer afterEach(t)
 	beforeEach(t)
-
+	isValidInterfaceNameAndState = func(ifName string) bool {
+		return true
+	}
 	rules := map[string][]infv1alpha1.IngressNodeFirewallRules{
 		fmt.Sprintf("%s0", interfacePrefix): {
 			{
@@ -154,7 +159,9 @@ func TestSyncInterfaceIngressRulesConnectAllowRule(t *testing.T) {
 func TestSyncInterfaceIngressRulesAttachAndDetachSingleInterface(t *testing.T) {
 	defer afterEach(t)
 	beforeEach(t)
-
+	isValidInterfaceNameAndState = func(ifName string) bool {
+		return true
+	}
 	rules := map[string][]infv1alpha1.IngressNodeFirewallRules{
 		fmt.Sprintf("%s0", interfacePrefix): {
 			{
@@ -220,7 +227,9 @@ func TestSyncInterfaceIngressRulesAttachAndDetachSingleInterface(t *testing.T) {
 func TestSyncInterfaceIngressRulesAttachAndDetach(t *testing.T) {
 	defer afterEach(t)
 	beforeEach(t)
-
+	isValidInterfaceNameAndState = func(ifName string) bool {
+		return true
+	}
 	rules := map[string][]infv1alpha1.IngressNodeFirewallRules{
 		fmt.Sprintf("%s0", interfacePrefix): {
 			{
@@ -304,7 +313,9 @@ func TestSyncInterfaceIngressRulesAttachAndDetach(t *testing.T) {
 func TestResyncInterfaceIngressRulesSingleInterface(t *testing.T) {
 	defer afterEach(t)
 	beforeEach(t)
-
+	isValidInterfaceNameAndState = func(ifName string) bool {
+		return true
+	}
 	rules := map[string][]infv1alpha1.IngressNodeFirewallRules{
 		fmt.Sprintf("%s0", interfacePrefix): {
 			{
@@ -359,7 +370,9 @@ func TestResyncInterfaceIngressRulesSingleInterface(t *testing.T) {
 func TestResyncInterfaceIngressRules(t *testing.T) {
 	defer afterEach(t)
 	beforeEach(t)
-
+	isValidInterfaceNameAndState = func(ifName string) bool {
+		return true
+	}
 	rules := map[string][]infv1alpha1.IngressNodeFirewallRules{
 		fmt.Sprintf("%s0", interfacePrefix): {
 			{

--- a/pkg/webhook/webhook.go
+++ b/pkg/webhook/webhook.go
@@ -10,6 +10,7 @@ import (
 	"github.com/openshift/ingress-node-firewall/pkg/failsaferules"
 	"github.com/openshift/ingress-node-firewall/pkg/utils"
 
+	"golang.org/x/sys/unix"
 	apierrors "k8s.io/apimachinery/pkg/api/errors"
 	"k8s.io/apimachinery/pkg/runtime"
 	"k8s.io/apimachinery/pkg/runtime/schema"
@@ -75,7 +76,35 @@ func validateIngressNodeFirewall(ctx context.Context, inf *ingressnodefwv1alpha1
 			schema.GroupKind{Group: ingressnodefwv1alpha1.GroupVersion.Group, Kind: ingressnodefwv1alpha1.IngressNodeFirewall{}.Kind},
 			inf.Name, allErrs)
 	}
+	if allErrs := validateINFInterfaces(ctx, inf.Spec.Interfaces, inf.Name, kubeClient); len(allErrs) > 0 {
+		return apierrors.NewInvalid(
+			schema.GroupKind{Group: ingressnodefwv1alpha1.GroupVersion.Group, Kind: ingressnodefwv1alpha1.IngressNodeFirewall{}.Kind},
+			inf.Name, allErrs)
+	}
 	return nil
+}
+
+func validateINFInterfaces(ctx context.Context, infInterfaces []string, infName string, kubeClient client.Client) field.ErrorList {
+	var allErrs field.ErrorList
+
+	for index, inf := range infInterfaces {
+		if inf == "" {
+			allErrs = append(allErrs,
+				field.Invalid(field.NewPath("Spec").Child("interfaces").Index(index),
+					infName, fmt.Sprintf("can not use blank interfae names")))
+		}
+		if len(inf) > unix.IFNAMSIZ {
+			allErrs = append(allErrs,
+				field.Invalid(field.NewPath("Spec").Child("interfaces").Index(index),
+					infName, fmt.Sprintf("interface %q is too long", inf)))
+		}
+		if inf[0] >= '0' && inf[0] <= '9' {
+			allErrs = append(allErrs,
+				field.Invalid(field.NewPath("Spec").Child("interfaces").Index(index),
+					infName, fmt.Sprintf("interface %q can't start with a number", inf)))
+		}
+	}
+	return allErrs
 }
 
 func validateINFRules(ctx context.Context, infRules []ingressnodefwv1alpha1.IngressNodeFirewallRules, infName string, kubeClient client.Client) field.ErrorList {

--- a/pkg/webhook/webhook_suite_test.go
+++ b/pkg/webhook/webhook_suite_test.go
@@ -145,6 +145,34 @@ var _ = BeforeSuite(func() {
 
 }, 60)
 
+var _ = Describe("Interfaces", func() {
+	var inf *ingressnodefwv1alpha1.IngressNodeFirewall
+
+	BeforeEach(func() {
+		inf = getIngressNodeFirewall("interfaces-validation")
+	})
+
+	Context("ingress node firewall interfaces validation", func() {
+		It("interfaces config with valid interfaces name", func() {
+			configInterfaces(inf, []string{"eth0", "eth1"})
+			Expect(createIngressNodeFirewall(inf)).To(Succeed())
+			Expect(deleteIngressNodeFirewall(inf)).To(Succeed())
+		})
+		It("interfaces config with empty strings", func() {
+			configInterfaces(inf, []string{""})
+			Expect(createIngressNodeFirewall(inf)).ToNot(Succeed())
+		})
+		It("interfaces config with too long interface name", func() {
+			configInterfaces(inf, []string{"abcd123459$%&ABDC44555555555"})
+			Expect(createIngressNodeFirewall(inf)).ToNot(Succeed())
+		})
+		It("interfaces config with interface name starts with number", func() {
+			configInterfaces(inf, []string{"0th"})
+			Expect(createIngressNodeFirewall(inf)).ToNot(Succeed())
+		})
+	})
+})
+
 var _ = Describe("Rules", func() {
 	Context("protocol is ICMPv4", func() {
 		var inf *ingressnodefwv1alpha1.IngressNodeFirewall
@@ -617,5 +645,11 @@ func getICMPv6Rule(order uint32, protocol ingressnodefwv1alpha1.IngressNodeFirew
 			},
 		},
 		Action: action,
+	}
+}
+
+func configInterfaces(inf *ingressnodefwv1alpha1.IngressNodeFirewall, intfs []string) {
+	for _, intf := range intfs {
+		inf.Spec.Interfaces = append(inf.Spec.Interfaces, intf)
 	}
 }

--- a/test/e2e/functional/tests/e2e.go
+++ b/test/e2e/functional/tests/e2e.go
@@ -204,23 +204,16 @@ var _ = Describe("Ingress Node Firewall", func() {
 				return true
 			}, infwutils.Timeout, infwutils.Interval).Should(BeTrue())
 
-			By("checking Ingress node firewall nodeState resource is deleted")
-			Eventually(func() bool {
-				for _, nState := range nodeStateList.Items {
-					if err := testclient.Client.Delete(context.Background(), &nState); err != nil {
-						if !errors.IsNotFound(err) {
-							return false
-						}
-					}
-				}
-				return true
-			}, infwutils.Timeout, infwutils.Interval).Should(BeTrue(), "Failed to delete IngressNodeFirewall custom resource")
-
 			By("checking Ingress node firewall rules resource is deleted")
 			Eventually(func() bool {
 				err := testclient.Client.Delete(context.Background(), rules)
 				return errors.IsNotFound(err)
 			}, infwutils.Timeout, infwutils.Interval).Should(BeTrue(), "Failed to delete IngressNodeFirewall custom resource")
+
+			By("checking Ingress node firewall nodeState resource is deleted when firewall rules object is deleted")
+			err = testclient.Client.List(context.Background(), nodeStateList)
+			Expect(err).ToNot(HaveOccurred())
+			Expect(len(nodeStateList.Items)).To(BeZero())
 		})
 	})
 
@@ -417,7 +410,6 @@ var _ = Describe("Ingress Node Firewall", func() {
 				Expect(err).ToNot(HaveOccurred())
 				Expect(daemonset.OwnerReferences).ToNot(BeNil())
 				Expect(daemonset.OwnerReferences[0].Kind).To(Equal("IngressNodeFirewallConfig"))
-
 				infwutils.DeleteINFConfig(config)
 			}
 		})
@@ -428,7 +420,7 @@ var _ = Describe("Ingress Node Firewall", func() {
 			infwutils.DefineWithInterface(inf, "eth0")
 			infwutils.DefineDenyTCPRule(inf, "1.1.1.1/32", 40000)
 			Expect(testclient.Client.Create(context.Background(), inf)).To(Succeed())
-			Expect(testclient.Client.Delete(context.Background(), inf)).To(Succeed())
+			cleanNodeFirewallRule(inf)
 		})
 
 		It("should allow valid ingressnodefirewall UDP rule", func() {
@@ -437,7 +429,7 @@ var _ = Describe("Ingress Node Firewall", func() {
 			infwutils.DefineWithInterface(inf, "eth0")
 			infwutils.DefineDenyUDPRule(inf, "1.1.1.1/32", 40000)
 			Expect(testclient.Client.Create(context.Background(), inf)).To(Succeed())
-			Expect(testclient.Client.Delete(context.Background(), inf)).To(Succeed())
+			cleanNodeFirewallRule(inf)
 		})
 
 		It("should allow valid ingressnodefirewall ICMPV4 rule", func() {
@@ -446,7 +438,7 @@ var _ = Describe("Ingress Node Firewall", func() {
 			infwutils.DefineWithInterface(inf, "eth0")
 			infwutils.DefineDenyICMPV4Rule(inf, "1.1.1.1/32")
 			Expect(testclient.Client.Create(context.Background(), inf)).To(Succeed())
-			Expect(testclient.Client.Delete(context.Background(), inf)).To(Succeed())
+			cleanNodeFirewallRule(inf)
 		})
 
 		It("should allow valid ingressnodefirewall ICMPV6 rule", func() {
@@ -455,7 +447,7 @@ var _ = Describe("Ingress Node Firewall", func() {
 			infwutils.DefineWithInterface(inf, "eth0")
 			infwutils.DefineDenyICMPV6Rule(inf, "1.1.1.1/32")
 			Expect(testclient.Client.Create(context.Background(), inf)).To(Succeed())
-			Expect(testclient.Client.Delete(context.Background(), inf)).To(Succeed())
+			cleanNodeFirewallRule(inf)
 		})
 
 		It("should allow valid ingressnodefirewall SCTP rule", func() {
@@ -464,7 +456,7 @@ var _ = Describe("Ingress Node Firewall", func() {
 			infwutils.DefineWithInterface(inf, "eth0")
 			infwutils.DefineDenySCTPRule(inf, "1.1.1.1/32", 40000)
 			Expect(testclient.Client.Create(context.Background(), inf)).To(Succeed())
-			Expect(testclient.Client.Delete(context.Background(), inf)).To(Succeed())
+			cleanNodeFirewallRule(inf)
 		})
 
 		It("should block any rules which conflict with failsafe rules", func() {
@@ -474,6 +466,7 @@ var _ = Describe("Ingress Node Firewall", func() {
 				infwutils.DefineWithInterface(inf, "eth0")
 				infwutils.DefineDenyTCPRule(inf, "1.1.1.1/32", tcpFailSafeRule.GetPort())
 				Expect(testclient.Client.Create(context.Background(), inf)).ToNot(Succeed())
+				cleanNodeFirewallRule(inf)
 			}
 			for _, udpFailSafeRule := range failsaferules.GetUDP() {
 				inf := infwutils.GetINF(OperatorNameSpace, fmt.Sprintf("e2e-webhook-block-conflict-%s-udp", udpFailSafeRule.GetServiceName()))
@@ -481,62 +474,20 @@ var _ = Describe("Ingress Node Firewall", func() {
 				infwutils.DefineWithInterface(inf, "eth0")
 				infwutils.DefineDenyUDPRule(inf, "1.1.1.1/32", udpFailSafeRule.GetPort())
 				Expect(testclient.Client.Create(context.Background(), inf)).ToNot(Succeed())
+				cleanNodeFirewallRule(inf)
 			}
-		})
-
-		It("should block invalid order", func() {
-			inf := infwutils.GetINF(OperatorNameSpace, "e2e-webhook-block-invalid-order")
-			infwutils.DefineWithWorkerNodeSelector(inf)
-			infwutils.DefineWithInterface(inf, "eth0")
-			infwutils.DefineDenyTCPRule(inf, "1.1.1.1/32", 40000)
-			// order zero is invalid
-			inf.Spec.Ingress[0].FirewallProtocolRules[0].Order = 0
-			Expect(testclient.Client.Create(context.Background(), inf)).ToNot(Succeed())
-		})
-
-		It("should block TCP rule ICMP defined", func() {
-			inf := infwutils.GetINF(OperatorNameSpace, "e2e-webhook-block-tcp-icmp")
-			infwutils.DefineWithWorkerNodeSelector(inf)
-			infwutils.DefineWithInterface(inf, "eth0")
-			infwutils.DefineDenyTCPRule(inf, "1.1.1.1/32", 40000)
-			inf.Spec.Ingress[0].FirewallProtocolRules[0].ProtocolConfig.ICMP = &ingressnodefwv1alpha1.IngressNodeFirewallICMPRule{ICMPCode: 8}
-			Expect(testclient.Client.Create(context.Background(), inf)).ToNot(Succeed())
-		})
-
-		It("should block UDP rule ICMP defined", func() {
-			inf := infwutils.GetINF(OperatorNameSpace, "e2e-webhook-block-udp-icmp")
-			infwutils.DefineWithWorkerNodeSelector(inf)
-			infwutils.DefineWithInterface(inf, "eth0")
-			infwutils.DefineDenyUDPRule(inf, "1.1.1.1/32", 40000)
-			inf.Spec.Ingress[0].FirewallProtocolRules[0].ProtocolConfig.ICMP = &ingressnodefwv1alpha1.IngressNodeFirewallICMPRule{ICMPCode: 8}
-			Expect(testclient.Client.Create(context.Background(), inf)).ToNot(Succeed())
-		})
-
-		It("should block SCTP rule ICMP defined", func() {
-			inf := infwutils.GetINF(OperatorNameSpace, "e2e-webhook-block-sctp-icmp")
-			infwutils.DefineWithWorkerNodeSelector(inf)
-			infwutils.DefineWithInterface(inf, "eth0")
-			infwutils.DefineDenySCTPRule(inf, "1.1.1.1/32", 40000)
-			inf.Spec.Ingress[0].FirewallProtocolRules[0].ProtocolConfig.ICMP = &ingressnodefwv1alpha1.IngressNodeFirewallICMPRule{ICMPCode: 8}
-			Expect(testclient.Client.Create(context.Background(), inf)).ToNot(Succeed())
-		})
-
-		It("should block ICMPV4 with TCP port defined", func() {
-			inf := infwutils.GetINF(OperatorNameSpace, "e2e-webhook-block-icmpv4-tcpport")
-			infwutils.DefineWithWorkerNodeSelector(inf)
-			infwutils.DefineWithInterface(inf, "eth0")
-			infwutils.DefineDenyICMPV4Rule(inf, "1.1.1.1/32")
-			inf.Spec.Ingress[0].FirewallProtocolRules[0].ProtocolConfig.TCP = &ingressnodefwv1alpha1.IngressNodeFirewallProtoRule{Ports: intstr.FromInt(80)}
-			Expect(testclient.Client.Create(context.Background(), inf)).ToNot(Succeed())
-		})
-
-		It("should block ICMPV6 with TCP port defined", func() {
-			inf := infwutils.GetINF(OperatorNameSpace, "e2e-webhook-block-icmpv6-tcpport")
-			infwutils.DefineWithWorkerNodeSelector(inf)
-			infwutils.DefineWithInterface(inf, "eth0")
-			infwutils.DefineDenyICMPV6Rule(inf, "1.1.1.1/32")
-			inf.Spec.Ingress[0].FirewallProtocolRules[0].ProtocolConfig.TCP = &ingressnodefwv1alpha1.IngressNodeFirewallProtoRule{Ports: intstr.FromInt(80)}
-			Expect(testclient.Client.Create(context.Background(), inf)).ToNot(Succeed())
 		})
 	})
 })
+
+func cleanNodeFirewallRule(inf *ingressnodefwv1alpha1.IngressNodeFirewall) {
+	nodeStateList := &ingressnodefwv1alpha1.IngressNodeFirewallNodeStateList{}
+
+	Eventually(func() bool {
+		err := testclient.Client.Delete(context.Background(), inf)
+		return errors.IsNotFound(err)
+	}, infwutils.Timeout, infwutils.Interval).Should(BeTrue())
+	err := testclient.Client.List(context.Background(), nodeStateList)
+	Expect(err).ToNot(HaveOccurred())
+	Expect(len(nodeStateList.Items)).To(BeZero())
+}

